File: packages/health_connector_hk_ios/test/unit_tests/src/mappers/metadata_mappers/recording_method_mapper_test.dart
import 'package:health_connector_core/health_connector_core_internal.dart';
import 'package:health_connector_hk_ios/src/mappers/metadata_mappers/recording_method_mapper.dart';
import 'package:health_connector_hk_ios/src/pigeon/health_connector_hk_ios_api.g.dart';
import 'package:parameterized_test/parameterized_test.dart';
import 'package:test/test.dart';

void main() {
  group(
    'RecordingMethodMapper',
    () {
      group(
        'RecordingMethodDtoMapper',
        () {
          parameterizedTest(
            'maps RecordingMethod to RecordingMethodDto',
            [
              [RecordingMethod.unknown, RecordingMethodDto.unknown],
              [RecordingMethod.manualEntry, RecordingMethodDto.manualEntry],
              [
                RecordingMethod.automaticallyRecorded,
                RecordingMethodDto.automaticallyRecorded,
              ],
              [
                RecordingMethod.activelyRecorded,
                RecordingMethodDto.activelyRecorded,
              ],
            ],
            (RecordingMethod method, RecordingMethodDto expectedDto) {
              final dto = method.toDto();
              expect(dto, expectedDto);
            },
          );
        },
      );

      group(
        'RecordingMethodDtoToDomain',
        () {
          parameterizedTest(
            'maps RecordingMethodDto to RecordingMethod',
            [
              [RecordingMethodDto.unknown, RecordingMethod.unknown],
              [RecordingMethodDto.manualEntry, RecordingMethod.manualEntry],
              [
                RecordingMethodDto.automaticallyRecorded,
                RecordingMethod.automaticallyRecorded,
              ],
              [
                RecordingMethodDto.activelyRecorded,
                RecordingMethod.activelyRecorded,
              ],
            ],
            (RecordingMethodDto dto, RecordingMethod expectedMethod) {
              final method = dto.toDomain();
              expect(method, expectedMethod);
            },
          );
        },
      );
    },
  );
}

File: packages/health_connector_hc_android/android/src/test/kotlin/com/phamtunglam/health_connector_hc_android/unit_tests/mappers/metadata_mappers/RecordingMethodMapperTest.kt
package com.phamtunglam.health_connector_hc_android.unit_tests.mappers.metadata_mappers

import androidx.health.connect.client.records.metadata.Metadata
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toRecordingMethodDto
import com.phamtunglam.health_connector_hc_android.pigeon.RecordingMethodDto
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource

/**
 * Unit tests for Recording Method Mapper.
 *
 * Tests verify proper mapping from Health Connect recording method constants to
 * [RecordingMethodDto], including edge case handling for unknown recording methods.
 */
@DisplayName("RecordingMethodMapper")
class RecordingMethodMapperTest {

    @Nested
    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
    @DisplayName("toRecordingMethodDto")
    inner class ToRecordingMethodDto {

        @ParameterizedTest(name = "constant={0} maps to dto={1}")
        @MethodSource("provideRecordingMethodMappings")
        @DisplayName(
            "GIVEN any recording method constant → " +
                "WHEN toRecordingMethodDto called → " +
                "THEN maps to correct RecordingMethodDto",
        )
        fun whenAnyRecordingMethod_thenMapsToCorrectDto(
            constant: Int,
            expectedDto: RecordingMethodDto,
        ) {
            // When
            val result = constant.toRecordingMethodDto()

            // Then
            result shouldBe expectedDto
        }

        fun provideRecordingMethodMappings(): List<Arguments> = listOf(
            // Valid recording method constants
            Arguments.of(
                Metadata.RECORDING_METHOD_MANUAL_ENTRY,
                RecordingMethodDto.MANUAL_ENTRY,
            ),
            Arguments.of(
                Metadata.RECORDING_METHOD_AUTOMATICALLY_RECORDED,
                RecordingMethodDto.AUTOMATICALLY_RECORDED,
            ),
            Arguments.of(
                Metadata.RECORDING_METHOD_ACTIVELY_RECORDED,
                RecordingMethodDto.ACTIVELY_RECORDED,
            ),
            Arguments.of(
                Metadata.RECORDING_METHOD_UNKNOWN,
                RecordingMethodDto.UNKNOWN,
            ),

            // Edge cases: Invalid/unknown constants should default to UNKNOWN
            Arguments.of(0, RecordingMethodDto.UNKNOWN),
            Arguments.of(-1, RecordingMethodDto.UNKNOWN),
            Arguments.of(-100, RecordingMethodDto.UNKNOWN),
            Arguments.of(100, RecordingMethodDto.UNKNOWN),
            Arguments.of(999, RecordingMethodDto.UNKNOWN),
            Arguments.of(9999, RecordingMethodDto.UNKNOWN),
            Arguments.of(Int.MAX_VALUE, RecordingMethodDto.UNKNOWN),
            Arguments.of(Int.MIN_VALUE, RecordingMethodDto.UNKNOWN),
        )
    }
}
